File: src/utils/transactionUtils.js
const crypto = require('crypto');
const secp256k1 = require('secp256k1');

function validateTransaction(transactionData) {
    // Additional checks

    // 1. Validate ScriptPubKey Address Formats
    for (const output of transactionData.vout) {
        if (output.scriptpubkey_type === "v1_p2tr" && !isValidBech32Address(output.scriptpubkey_address, "bc")) {
            return false;
        } else if (output.scriptpubkey_type === "p2sh" && !isValidBase58Address(output.scriptpubkey_address, "bc")) {
            return false;
        }
    }

    // 2. Check Transaction Fee
    const totalOutputValue = transactionData.vout.reduce((acc, output) => acc + output.value, 0);
    const totalInputValue = transactionData.vin.reduce((acc, input) => {
        const prevOutput = input.prevout;
        return acc + (prevOutput ? prevOutput.value : 0);
    }, 0);
    const transactionFee = totalInputValue - totalOutputValue;
    if (transactionFee < 0) {
        return false; // Negative fee, invalid transaction
    }

    // 3. Confirm Coinbase Transaction Validation
    for (const input of transactionData.vin) {
        if (input.is_coinbase && input.prevout.value !== 0) {
            return false; // Invalid coinbase transaction
        }
    }

    // 4. Verify Locktime (if present)
    if (transactionData.locktime && transactionData.locktime < 0) {
        return false; // Invalid locktime
    }

    // 5. Check for Negative Values
    if (totalOutputValue < 0 || totalInputValue < 0) {
        return false; // Negative values, invalid transaction
    }

    // 6. Validate Witness Data
    for (const input of transactionData.vin) {
        if (!input.is_coinbase && (!input.witness || input.witness.length === 0)) {
            return false; // Witness data missing for non-coinbase transaction
        }
    }

    // 7. Remove Dust Transactions
    const dustThreshold = 546; // Assuming a dust threshold of 546 satoshis
    for (const output of transactionData.vout) {
        if (output.value < dustThreshold) {
            return false; // Dust output, invalid transaction
        }
    }

    // 8. Check for Double Spending
    const spentOutputs = new Set();
    for (const input of transactionData.vin) {
        const outputHash = `${input.txid}:${input.vout}`;
        if (spentOutputs.has(outputHash)) {
            return false; // Double spending detected, invalid transaction
        }
        spentOutputs.add(outputHash);
    }

    // 9. Validate Script Formats
    for (const output of transactionData.vout) {
        const { scriptpubkey, scriptpubkey_type } = output;

        if (scriptpubkey_type === "p2sh") {
            // Validate P2SH script format
            const pattern = /^OP_HASH160 ([0-9a-fA-F]{40}) OP_EQUAL$/;
            const match = scriptpubkey.match(pattern);
            if (!match) {
                return false; // Invalid P2SH script format
            }

            const addressHash = match[1];
            const address = getAddressFromHash(addressHash, "p2sh");
            if (address !== output.scriptpubkey_address) {
                return false; // Address mismatch for P2SH
            }
        } else if (scriptpubkey_type === "v1_p2tr") {
            // Validate P2TR script format
            const pattern = /^OP_PUSHNUM_1 OP_PUSHBYTES_32 ([0-9a-fA-F]{64})$/;
            const match = scriptpubkey.match(pattern);
            if (!match) {
                return false; // Invalid P2TR script format
            }

            const witnessHash = match[1];
            const address = getAddressFromHash(witnessHash, "p2tr");
            if (address !== output.scriptpubkey_address) {
                return false; // Address mismatch for P2TR
            }
        }
        // Add more script format validations as needed
    }

    // 10. Extract Signatures and Perform ECDSA Verification
    for (const input of transactionData.vin) {
        if (!input.is_coinbase) {
            const { witness, prevout } = input;
            const { scriptpubkey, value } = prevout;

            if (scriptpubkey.startsWith("OP_DUP OP_HASH160")) {
                // P2PKH transaction
                const signature = witness[0];
                const pubkey = witness[1];

                // Perform ECDSA signature verification
                if (!verifySignature(signature, prevout, pubkey)) {
                    return false; // Invalid signature for P2PKH
                }
            } else if (scriptpubkey.startsWith("OP_PUSHNUM_1 OP_PUSHBYTES_32")) {
                // P2TR transaction
                const witnessScript = witness[witness.length - 1];

                // Perform ECDSA signature verification for witness script
                if (!verifyWitnessScript(witnessScript, prevout, witness)) {
                    return false; // Invalid witness script for P2TR
                }
            }
        }
    }

    // 11. Maximize Transaction Fee
    const sortedOutputs = transactionData.vout.sort((a, b) => b.value - a.value);
    const maximizedTransactionData = {
        ...transactionData,
        vout: sortedOutputs
    };

    // Original validation checks
    // Check if the outputs are valid
    let totalOutput = 0;
    for (const output of maximizedTransactionData.vout) {
        // Validate output values
        if (!output.scriptpubkey || !output.value || output.value < 0) {
            return false;
        }
        totalOutput += output.value;
    }

    // Check if the transaction version is valid
    const validVersion = [2]; // Add more valid versions as needed
    if (!validVersion.includes(maximizedTransactionData.version)) {
        return false;
    }

    // Calculate the total input value
    let totalInput = 0;
    for (const input of maximizedTransactionData.vin) {
        // Validate input values
        if (!input.txid || input.vout === undefined || input.vout < 0) {
            return false;
        }

        // Get the value of the previous output being spent
        const prevOutput = input.prevout;
        if (!prevOutput || !prevOutput.value || prevOutput.value < 0) {
            return false;
        }
        totalInput += prevOutput.value;

        // Check if the input is a coinbase transaction and validate accordingly
        if (input.is_coinbase) {
            // Perform additional validation for coinbase transactions if needed
        }
    }

    // Check if the total output value is valid
    // The total output value should not be greater than the total input value
    if (totalOutput > totalInput) {
        return false;
    }

    // If all validation checks pass, return true
    return true;
}

// Helper functions for address and signature verification
function isValidBech32Address(address, prefix) {
    // Implement Bech32 address validation logic using permissible libraries
    // Return true if the address is valid, false otherwise
    return true; // Placeholder
}

function isValidBase58Address(address, prefix) {
    // Implement Base58 address validation logic using permissible libraries
    // Return true if the address is valid, false otherwise
    return true; // Placeholder
}

function verifySignature(signature, prevout, pubkey) {
    const { scriptpubkey, value } = prevout;
    const message = getSignatureMessage(scriptpubkey, value);
    const signatureBuffer = Buffer.from(signature, 'hex');
    const isValid = secp256k1.ecdsaVerify(signatureBuffer, message);
    return isValid;
}

function verifyWitnessScript(witnessScript, prevout, witness) {
    // Implement witness script validation logic here
    // This will depend on the specific witness script format and signature schemes used
    return true; // Placeholder for now
}

function getSignatureMessage(scriptpubkey, value) {
    // Implement signature message generation logic here
    // This will depend on the specific transaction and script format
    return ''; // Placeholder for now
}

module.exports = { validateTransaction };

File: src/utils/transactionUtils.py

import hashlib
import secp256k1
from typing import List, Dict
import re



def validate_transaction(transaction_data: Dict) -> bool:
    # 1. Validate ScriptPubKey Address Formats
    for output in transaction_data["vout"]:
        if output["scriptpubkey_type"] == "v1_p2tr" and not is_valid_bech32_address(
            output["scriptpubkey_address"], "bc"
        ):
            return False
        elif output["scriptpubkey_type"] == "p2sh" and not is_valid_base58_address(
            output["scriptpubkey_address"], "bc"
        ):
            return False

    # 2. Check Transaction Fee
    total_output_value = sum(output["value"] for output in transaction_data["vout"])
    total_input_value = sum(
        input["prevout"]["value"]
        for input in transaction_data["vin"]
        if input["prevout"]
    )
    transaction_fee = total_input_value - total_output_value
    if transaction_fee < 0:
        return False

    # 3. Confirm Coinbase Transaction Validation
    for input in transaction_data["vin"]:
        if input["is_coinbase"] and input["prevout"]["value"] != 0:
            return False

    # 4. Verify Locktime (if present)
    if "locktime" in transaction_data and transaction_data["locktime"] < 0:
        return False

    # 5. Check for Negative Values
    if total_output_value < 0 or total_input_value < 0:
        return False

    # 6. Validate Witness Data
    for input in transaction_data["vin"]:
        if not input["is_coinbase"] and (
            "witness" not in input or not input["witness"] or len(input["witness"]) == 0
        ):
            return False

    # 7. Remove Dust Transactions
    dust_threshold = 546  # Assuming a dust threshold of 546 satoshis
    for output in transaction_data["vout"]:
        if output["value"] < dust_threshold:
            return False

    # 8. Check for Double Spending
    spent_outputs = set()
    for input in transaction_data["vin"]:
        output_hash = f"{input['txid']}:{input['vout']}"
        if output_hash in spent_outputs:
            return False
        spent_outputs.add(output_hash)

    # 9. Validate Script Formats
    for output in transaction_data["vout"]:
        scriptpubkey = output["scriptpubkey"]
        scriptpubkey_type = output["scriptpubkey_type"]

        if scriptpubkey_type == "p2sh":
            # Validate P2SH script format
            pattern = r"^OP_HASH160 ([0-9a-fA-F]{40}) OP_EQUAL$"
            match = re.match(pattern, scriptpubkey)
            if not match:
                return False

            address_hash = match.group(1)
            address = get_address_from_hash(address_hash, "p2sh")
            if address != output["scriptpubkey_address"]:
                return False

        elif scriptpubkey_type == "v1_p2tr":
            # Validate P2TR script format
            pattern = r"^OP_PUSHNUM_1 OP_PUSHBYTES_32 ([0-9a-fA-F]{64})$"
            match = re.match(pattern, scriptpubkey)
            if not match:
                return False

            witness_hash = match.group(1)
            address = get_address_from_hash(witness_hash, "p2tr")
            if address != output["scriptpubkey_address"]:
                return False

    # 10. Extract Signatures and Perform ECDSA Verification
    for input in transaction_data["vin"]:
        if not input["is_coinbase"]:
            witness = input["witness"]
            prevout = input["prevout"]
            scriptpubkey = prevout["scriptpubkey"]
            value = prevout["value"]

            if scriptpubkey.startswith("OP_DUP OP_HASH160"):
                # P2PKH transaction
                signature = witness[0]
                pubkey = witness[1]

                # Perform ECDSA signature verification
                if not verify_signature(signature, prevout, pubkey):
                    return False

            elif scriptpubkey.startswith("OP_PUSHNUM_1 OP_PUSHBYTES_32"):
                # P2TR transaction
                witness_script = witness[-1]

                # Perform ECDSA signature verification for witness script
                if not verify_witness_script(witness_script, prevout, witness):
                    return False

    # 11. Maximize Transaction Fee
    sorted_outputs = sorted(
        transaction_data["vout"], key=lambda x: x["value"], reverse=True
    )
    maximized_transaction_data = {**transaction_data, "vout": sorted_outputs}

    # Original validation checks
    # Check if the outputs are valid
    total_output = sum(output["value"] for output in maximized_transaction_data["vout"])
    for output in maximized_transaction_data["vout"]:
        if not output["scriptpubkey"] or not output["value"] or output["value"] < 0:
            return False

    # Check if the transaction version is valid
    valid_version = [2]  # Add more valid versions as needed
    if transaction_data["version"] not in valid_version:
        return False

    # Calculate the total input value
    total_input = 0
    for input in maximized_transaction_data["vin"]:
        if not input["txid"] or input["vout"] is None or input["vout"] < 0:
            return False

        # Get the value of the previous output being spent
        prev_output = input["prevout"]
        if not prev_output or not prev_output["value"] or prev_output["value"] < 0:
            return False
        total_input += prev_output["value"]

        # Check if the input is a coinbase transaction and validate accordingly
        if input["is_coinbase"]:
            # Perform additional validation for coinbase transactions if needed
            pass

    # Check if the total output value is valid
    # The total output value should not be greater than the total input value
    if total_output > total_input:
        return False

    # If all validation checks pass, return true
    return True


# Helper functions for address and signature verification
def is_valid_bech32_address(address: str, prefix: str) -> bool:
    # Implement Bech32 address validation logic using permissible libraries
    # Return true if the address is valid, false otherwise
    return True  # Placeholder


def is_valid_base58_address(address: str, prefix: str) -> bool:
    # Implement Base58 address validation logic using permissible libraries
    # Return true if the address is valid, false otherwise
    return True  # Placeholder


def verify_signature(signature: str, prevout: Dict, pubkey: str) -> bool:
    scriptpubkey = prevout["scriptpubkey"]
    value = prevout["value"]
    message = get_signature_message(scriptpubkey, value)
    signature_buffer = bytes.fromhex(signature)
    is_valid = secp256k1.ecdsa_verify(signature_buffer, message)
    return is_valid


def verify_witness_script(
    witness_script: str, prevout: Dict, witness: List[str]
) -> bool:
    # Implement witness script validation logic here
    # This will depend on the specific witness script format and signature schemes used
    return True  # Placeholder for now


def get_signature_message(scriptpubkey: str, value: int) -> str:
    # Implement signature message generation logic here
    # This will depend on the specific transaction and script format
    return ""  # Placeholder for now


def get_address_from_hash(hash: str, address_type: str) -> str:
    if address_type == "p2sh":
        # Placeholder implementation for P2SH address generation
        # Example: converting hash to base58 address format
        return "P2SH_ADDRESS_PLACEHOLDER"
    elif address_type == "p2tr":
        # Placeholder implementation for P2TR address generation
        # Example: converting hash to Bech32 address format
        return "P2TR_ADDRESS_PLACEHOLDER"
    else:
        raise ValueError("Unsupported address type")


if __name__ == "__main__":
    # Example usage
    transaction_data = {
        # Insert transaction data here
    }
    is_valid = validate_transaction(transaction_data)
    print("Transaction is valid:", is_valid)
